src/metabase/models/setting.clj

```diff
@@ -84,12 +90,109 @@
 ;; Cache is a 1:1 mapping of what's in the DB
 ;; Cached lookup time is ~60µs, compared to ~1800µs for DB lookup
 
-(defonce ^:private cache
+(def ^:private cache
+  "Settings cache. Map of Setting key (string) -> Setting value (string)."
   (atom nil))
 
-(defn- restore-cache-if-needed! []
-  (when-not @cache
-    (reset! cache (db/select-field->field :key :value Setting))))
+;; CACHE SYNCHRONIZATION
+;;
+;; When running multiple Metabase instances (horizontal scaling), it is of course possible for one instance to update
+;; a Setting, and, since Settings are cached (to avoid tons of DB calls), for the other instances to then have an
+;; out-of-date cache. Thus we need a way for instances to know when their caches are out of date, so they can update
+;; them accordingly. Here is our solution:
+;;
+;; We will record the last time *any* Setting was updated in a special Setting called `settings-last-updated`.
+;;
+;; Since `settings-last-updated` itself is a Setting, it will get fetched as part of each instance's local cache; we
+;; can then periodically compare the locally cached value of `settings-last-updated` with the value in the DB. If our
+;; locally cached value is older than the one in the DB, we will flush our cache. When the cache is fetched again, it
+;; will have the up-to-date value.
+;;
+;; Because different machines can have out-of-sync clocks, we'll rely entirely on the application DB for caclulating
+;; and comparing values of `settings-last-updated`. Because the Setting table itself only stores text values, we'll
+;; need to cast it between TEXT and TIMESTAMP SQL types as needed.
+
+(def ^:private ^String settings-last-updated-key "settings-last-updated")
+
+(defn- update-settings-last-updated!
+  "Update the value of `settings-last-updated` in the DB; if the row does not exist, insert one."
+  []
+  (log/debug (trs "Updating value of settings-last-updated in DB..."))
+  ;; attempt to UPDATE the existing row. If no row exists, `update-where!` will return false...
+  (or (db/update-where! Setting {:key settings-last-updated-key} :value (hx/cast :text (hsql/raw "current_timestamp")))
+      ;; ...at which point we will try to INSERT a new row. Note that it is entirely possible two instances can both
+      ;; try to INSERT it at the same time; one instance would fail because it would violate the PK constraint on
+      ;; `key`, and throw a SQLException. As long as one instance updates the value, we are fine, so we can go ahead
+      ;; and ignore that Exception if one is thrown.
+      (try
+        (db/insert! Setting :key settings-last-updated-key, :value (hx/cast :text (hsql/raw "current_timestamp")))
+        (catch java.sql.SQLException _)))
```

Comment:
Couldn't there be other reasons for a SQLException here? It seems like it would be useful to log it as info or something with a related message just in case this would swallow something that would be useful in debugging

Reply:
good point I will tweak